dowhy/causal_refuters/dummy_outcome_refuter.py

```diff
@@ -98,59 +103,87 @@ def refute_estimate(self):
 
         sample_estimates = np.zeros(self._num_simulations)
         self.logger.info("Refutation over {} simulated datasets".format(self._num_simulations) )
-        self.logger.info("The transformation passed: {}", self._transformations)
+        self.logger.info("The transformation passed: {}".format(self._transformations) )
+
+        data_chunks = self.data_preprocess()
+        estimates = []
+        for chunk in data_chunks:
+
+            X_input = chunk[self._chosen_variables]
+            new_outcome = chunk['y']
+            X = self._data[self._chosen_variables]
 
-        # This flag is to make sure we store the estimators whose input is deterministic
-        save_estimators = True
-        # We store the value of the estimators in the format "estimator_name" +  "pos_in_transform" : estimator_object
-        saved_estimator_dict = {}
-        
-        X = self._data[self._chosen_variables]
-        new_outcome = self._data['y']
-        
-        for index in range(self._num_simulations):
-            transform_num = 0
             for action, func_args in self._transformations:
 
                 if callable(action):
-                    new_outcome = action(X, **func_args)
-
+                    estimator = action(X_input, new_outcome, **func_args)
+                    new_outcome = estimator(X)
                 elif action in DummyOutcomeRefuter.SUPPORTED_ESTIMATORS:
-                    if action + str(transform_num) in saved_estimator_dict:
-                        estimator = saved_estimator_dict[action + str(transform_num)]
-                        new_outcome = estimator(X)
-                    else:
-                        estimator = self._estimate_dummy_outcome(func_args, action, new_outcome)
-                        new_outcome = estimator(X)
-                        if save_estimators:
-                            saved_estimator_dict[action + str(transform_num)] = estimator
-
+                    estimator = self._estimate_dummy_outcome(func_args, action, new_outcome, X_input)
+                    new_outcome = estimator(X)
                 elif action == 'noise':
-                    save_estimators = False
                     new_outcome = self._noise(new_outcome, func_args)
-
                 elif action == 'permute':
-                    save_estimators = False
                     new_outcome = self._permute(new_outcome, func_args)
-
                 elif action =='zero':
-                    save_estimators = False
                     new_outcome = np.zeros(new_outcome.shape)
+
+            new_data = chunk.assign(dummy_outcome=new_outcome)

```

Comment:
looks like the evaluation is being done on the same chunk. But as we discussed, good to do it on the other chunks. So chunk_i's dummy outcome generating model is used for setting chunk_j's outcome and then treatment effect is estimated on chunk_j. Similarly, you can test on all chunk_k such that k !=i

Reply:
Agreed, so I need to be able to run the data on the complement of the data taken into consideration.